### PR TITLE
[FEAT] Flexible bulk seed selling

### PR DESCRIPTION
**Pull request by Thimey**
# Description

Add the option to sell custom amount of crops.

"Custom" button added below "Sell All" which allows farmers to use a range of increment/decrement buttons to adjust the amount they desire to sell. This button will show when farmer has > 10 crops to sell.

Addresses #693 

### Before

image

### After

bulkSell5

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

 - Manual test

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by JCtapuk** [comment]
This functionality is convenient for the PC but not from a smartphone. I recommend adding a small button with a handle icon next to the sale on the right. When clicked, it will open a modal window in which you can click on 9 variations of the number buttons as well as 3 variations of the 25%, 50% and 100% buttons. This will give players a convenient way to specify the right amount or just click 50% and take if you have 200 pieces in your inventory then the number innput form will immediately indicate 100 pieces.
And 1 button is to clear the last character on the right if you made a mistake selecting quantities.

**Comment by alvinphoebeartemis** [comment]
Why not like this? 😊 I think this is much better.

**Comment by Thimey** [comment]
@JCtapuk, yep good point! I'll have a play around with your idea and also @alvinphoebeartemis suggestion to see what works best

**Comment by Z1kou** [comment]
I believe this feature has already been discussed by the devs and team a few weeks ago, the opinion's were different and at the end we settled to keep it as it is since we havent gotten a modal that is both simple and efficient both on PC and mobile !

**Comment by alvinphoebeartemis** [comment]
Hi Team, @Z1kou @Thimey 

How about this UI/UX. 🤗
image

**Comment by adamhannigan** [comment]
@Thimey Hey mate great initiative 👏 

We've faced this issue a few times before and found the slider approach isn't very user friendly on the mobile. Where possible we take the button/click approach as it is intuitive on mobile especially.

Exploring some of the button approaches above seems like a good start 💪

**Comment by Thimey** [comment]
Thanks @adamhannigan / @Z1kou - noted!

Nice work @alvinphoebeartemis! I'll play around with this - As Adam suggested, I'll try and come up with a solution with just buttons. 

A simple solution could be to add some +-1 and +- 10 buttons in the confirmation dialog and renaming the button to "Sell More" (I'm sure theres a better name for this, but you get the idea). This keeps the flow the same for selling all, but allows users to adjust before confirming.

I'll whip something up in the next couple of days 👍

**Comment by Thimey** [comment]
@adamhannigan / @Z1kou - Updated based on feedback

Designed inspired by @alvinphoebeartemis 🙂 

I'll tidy up code once we're all  happy with UX.

**Comment by spencerdezartsmith** [comment]
> Hi @Thimey 😊
> 
> I'm very impressed with the idea of pop-up when clicked the "Sell All" with your layout. image It gives me an idea for this mockup calculator-like. 😎
> 
> You may choose which one is better. **Mockup No. 1** image
> 
> **Mockup No. 2** image
> 
> Sir @adamhannigan can also check on this 😉

I this the mock-up 1 is better suited as it will work well on mobile as well. It would be good to have this a reusable component that is particularly tied to the buy/sell flow.

**Comment by Thimey** [comment]
oooh nice @alvinphoebeartemis!

However, I am wondering how often people would explicitly enter an amount to sell as opposed to incrementing/decrementing from total stock. It would be good to get some validation from others if this would save them time vs having a variety of different increment/decrement button +-1, +=10, +=20, MIN, MAX.

Adding explicit number entering would mean we need to think about overflow. Can they enter any number and stock overflow disables button and has some "invalid state" OR should they never be able to enter an amount greater than stock.

If we go with calculator style, we should at least include decrement buttons too.

Happy to implement either solution, but be great to get some more validation before committing 🙂

**Comment by adamhannigan** [comment]
Great work so far 👏 

Just to add some thoughts. How often do we think that a user would want to sell a custom amount?

The typically workflow right now is plant -> harvest -> sell. For majority of users I predict they would click the 'Sell All' button.

Adding an extra step would add friction to this flow. I would suggest ensuring that the workflow is firstly suited to accomodate the majority of users and how they play. It should be the 5-10% who want to do a custom bulk sell that need to do an extra step.

**Comment by Thimey** [comment]
Yep totally agree @adamhannigan. If we default to All and keep the option to edit in the confirmation dialog, there is no change to current "sell all" flow (2 clicks).

Since it is a less common flow to adjust the amount. I am leaning towards a collection of increment/decrement buttons over full calculator. This would be a simpler iteration and we can always change later to a calculator if it's needed.

**Comment by alvinphoebeartemis** [comment]
Hi @Thimey , the default value will still the max available number of crops. To prevent hurting the User Experience. 🤗 No additional process at all.

**Comment by Thimey** [comment]
Hey team, made a few changes, but decided not to go full calculator as I believe this is overkill and would be more clunky to use in a lot of situations.

 - Render increment buttons 1, 10, 50 depending on stock amount
 - Made the layout similar to @alvinphoebeartemis designs above, but with simpler buttons

As mentioned previously, there is no change to the flow of you want to sell all -> still 2 clicks with no changes.

**I'll update PR description once everyone is happy with the design**

bulkSell4

### Variations

**<= 10**
image
**<= 50**
image
**> 50**
image

**Comment by Pranav-yadav** [comment]
Great idea's and implementations here so far.. 🚀
@Thimey I think we should separate the Custom Sell functionality from Sell All as, most of the players will be using Sell All directly and showing them Custom Functionality wouldn't be good, UX.. As, @adamhannigan has mentioned..

My suggestion:
Why not keep a another button named **Custom** for custom functionality, besides Sell All, this will improve the UX.. 😉

Let me know what you guys think..

**Comment by Thimey** [comment]
Yeah nice one, @Pranav-yadav - makes sense to me. I'll update shortly

**Comment by Thimey** [comment]
Fellow farmers 🧑‍🌾 

I've addressed the UX concerns and update the PR description so this should be reviewable now 👍

**Review by Pranav-yadav — CHANGES_REQUESTED** [review]

File: src/features/crops/components/Plants.tsx
Comment (Thimey) [reply src/features/crops/components/Plants.tsx:130]: > Why not keep it < 5 ?

No big reason, just thought it wouldn't hurt (Technically less clicks to sell 4 😉). Happy to change though if we think it won't confuse people wondering why it's disabled.

> We should also focus on more clutter free UX for quick interactions.
> @Rofrtd would agree on this..

Are you suggesting another change or is this referring to the change mentioned above?

Comment (Pranav-yadav) [reply src/features/crops/components/Plants.tsx:130]: Yep. Actually what I was trying to say is that we should hide the Custom button for above mentioned conditions..

For more clutter free UI/UX.. 👍

Comment (Thimey) [reply src/features/crops/components/Plants.tsx:130]: Yep, was thinking the same thing... I think it's also worth checking for > 10 since anything less is still reasonable to sell single at time.

Comment (Pranav-yadav) [reply src/features/crops/components/Plants.tsx]: Definitely..
 `< 5` cause some crops cost higher (compared to sunflower xD) e.g. Wheat and Radish.. 
 the poor (noob) farmers should also be able to see the custom button. just taste of it.. 😉

**Comment by TeissierYannis** [comment]
Hello, what do you think about that design #844 
I do a duplicate pr sorry for that boys

**Comment by alvinphoebeartemis** [comment]
Hi @TeissierYannis thank you for bringing-in the idea. But for now we refrain from using manual input of numbers using inputbox. 😊 We suggest to have a ingame button to improve the UX. @Thimey has already a patch for this and we are waiting for the approval of the community. 😊 You may check our previous conversation on this PR. Thank you.

**Comment by adamhannigan** [comment]
@Thimey Love the initiative mate 👏 

For now we are focussing our attention on a few of the upcoming game features and are putting a hold on large UI features getting added. This is to prevent application bloat and less items the team will need to maintain. The aim is to try free up as much time over the next 2-3 months so we can launch some amazing features 🤗

**Pull request closed without merging**
